### ISSUE TITLE
[java] The ClassNamingConventionsRule false-positive's on the class name "Constants"

### ISSUE THREAD

**Issue by scottashipp** [pmd/pmd#3563]
**Affects PMD Version:** 
6.29.0

**Rule:**

ClassNamingConventions

Please provide the rule name and a link to the rule documentation:

https://pmd.github.io/latest/pmd_rules_java_codestyle.html#classnamingconventions

**Description:**

The class name `Constants` triggers a P1 warning of the ClassNamingConventions rule. From the documentation, the regular expression should be `[a-z0-9A-Z]*` (notice the star which means "0 or more") but in the code, it uses `[a-z0-9A-Z]+` (notice the plus, which means 1 or more). 

See https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/ClassNamingConventionsRule.java#L33 where the "+" is defined.

Because of the "+", it effectively requires that there be some prefix in the name. For example, `FooConstants`, or `UhConstants` would be valid but `Constants` is not. 

Given the mismatch between the [public documentation](https://pmd.github.io/latest/pmd_rules_java_codestyle.html#classnamingconventions) and the [code](https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/ClassNamingConventionsRule.java#L33), I believe this is unintentional.

**Code Sample demonstrating the issue:**

```
public class Constants {
   // ...
}
```

**Expected outcome:**

PMD reports a violation, but that's wrong. That's a false positive as far as I can see.

**Running PMD through:** Maven

**Comment by adangel** [comment]
Hm... your class `Constants` is obviously seen as a utility class instead of a "regular" class. Hence the pattern for utility classes is applied (`[A-Z][a-zA-Z0-9]+(Utils?|Helper|Constants)`).

I guess, you Constants class is indeed a utility class (e.g. has no non-static members).

> Given the mismatch between the public documentation and the code, I believe this is unintentional.

This can't happen - this part of the documentation is generated from the code - so it always matches...

The question is more like - should we allow the very generic name "Constants" as a name of a utility class. The question the comes up is "What constants?" - Constants can be everything, so not sure if it is a useful name...

By the way: You can configure the rule yourself: [Configuring rules](https://pmd.github.io/latest/pmd_userdocs_configuring_rules.html). Just set the regex you want.

This issue would be about changing the default regex - and I'm not sure, we should do this.

**Comment by scottashipp** [comment]
@adangel TBH I don't agree with the question you have stated. That isn't my contention and I didn't file this to debate a question like that. If I did, I'd be willing to make the argument that now that we have modules in Java 9 and packages can better encapsulate related concepts, the name `Constants` for a class would actually make sense in some contexts considering that the package it is in already categorizes it.

But that's not my argument nor what I'm here for. The reason I filed the ticket is that this rule triggers a violation on a case it obviously wasn't intended to, which is clearly an unintentional side effect. I looked back at the history of the rule before filing. It is based on a tweet from Kevlin Henney about classes named _____Helper or ____Utils. I am not sure how or when "Constants" got lumped into this but there's certainly nothing about disallowing the class name `Constants` in the rule history or in the Java ecosystem in general. The closest you can say is that a large class of global constants (whatever it's name) is considered an anti-pattern. (Or, but this isn't that close, an interface that holds constants.)

But this is the class naming convention rule. There are other rules for those particular anti-patterns.

**Comment by adangel** [comment]
Thanks for taking time to look into this.

> If a rule has a false positive then I think we all agree it should be changed.

Agreed.

Whether that's a false positive or not depends on what the rule should do. You are correct in questioning, whether this rule does more than it should.

ClassNamingConventions is one of the area, where endless bike-shed discussions are possible. Hence the rule is very configurable, because usually the default settings don't fit everybody's need.
E.g. - if you don't want to use the (probably not perfect) utility class detection, you can set this property to the same pattern like the class name pattern:

```xml
<rule ref="category/java/codestyle.xml/ClassNamingConventions">
    <properties>
        <property name="utilityClassPattern" value="[A-Z][a-zA-Z0-9]*" />
    </properties>
</rule>
```

This effectively allows "Constants" as a valid class name (Disclaimer: I didn't try this).

Please also note, that you are not the first person questioning this: e.g. #1595 #1096
There is also the request to increase the responsibility of this rule and add a testClassPattern: #2867 

One way forward would be: Make the ClassNamingConventionsRule simple - don't care about utility classes or anything other special. Basically, what you are proposing.
A middle ground would be, to use the same pattern for utilityClasses as for "normal" classes by default - if someone wants to handle utilityClasses differently, then it would be possible to configure the rule. We still have to describe, what PMD thinks, a utility class is (#1102).
On the other end of the possibilities is: Leave this rule as is - you need to override the default configuration, if you don't want to use the utility detection.

I have no strong opinion on this. Whatever we decide, there will always be someone who is unhappy with the decision.

**Comment by scottashipp** [comment]
@adangel Thank you! This change helps a lot! It really makes a difference!
